pkg/api/v1alpha1/cluster_webhook.go

```diff
@@ -377,7 +382,157 @@ func ValidateKubernetesVersionSkew(new, old *Cluster) field.ErrorList {
+	oldTopKubeVersion := old.Spec.KubernetesVersion
+
+	workerNodeGroupMap := make(map[string]*WorkerNodeGroupConfiguration)
+	for i := range old.Spec.WorkerNodeGroupConfigurations {
+		workerNodeGroupMap[old.Spec.WorkerNodeGroupConfigurations[i].Name] = &old.Spec.WorkerNodeGroupConfigurations[i]
+	}
+	for _, nodeGroupNewSpec := range new.Spec.WorkerNodeGroupConfigurations {
+		if workerNodeGrpOldSpec, ok := workerNodeGroupMap[nodeGroupNewSpec.Name]; ok {
+			oldVersion := workerNodeGrpOldSpec.KubernetesVersion
+			newVersion := nodeGroupNewSpec.KubernetesVersion
+
+			allErrs = performWorkerKubernetesValidations(oldVersion, newVersion, oldTopKubeVersion, topKubeVersion)
+		}
+	}
+
+	return allErrs
+}
+
+func performWorkerKubernetesValidations(oldVersion, newVersion *KubernetesVersion, oldTopKubeVersion, topKubeVersion KubernetesVersion) field.ErrorList {
+	var allErrs field.ErrorList
+	path := field.NewPath("spec").Child("WorkerNodeConfiguration.kubernetesVersion")
+
+	if oldVersion != nil && newVersion == nil {
+		allErrs = append(
+			allErrs,
+			validateRemoveWorkerKubernetesVersion(topKubeVersion, oldTopKubeVersion, newVersion, oldVersion)...,
+		)
+	}
+
+	if oldVersion != nil && newVersion != nil {
+		allErrs = append(
+			allErrs,
+			validateKubeVersionSkew(*newVersion, *oldVersion, path)...,
+		)
+	}
+
+	if oldVersion == nil && newVersion != nil {
+		allErrs = append(
+			allErrs,
+			validateKubeVersionSkew(*newVersion, oldTopKubeVersion, path)...,
+		)
+	}
+
+	if newVersion != nil {
+		allErrs = append(
+			allErrs,
+			validateCPWorkerKubeSkew(topKubeVersion, *newVersion)...,
+		)
+	}
+
+	return allErrs
+}
+
+func validateRemoveWorkerKubernetesVersion(newCPVersion, oldCPVersion KubernetesVersion, newWorkerVersion, oldWorkerVersion *KubernetesVersion) field.ErrorList {
+	var allErrs field.ErrorList
+	path := field.NewPath("spec").Child("WorkerNodeConfiguration.kubernetesVersion")
+
+	parsedWorkerVersion, err := version.ParseGeneric(string(*oldWorkerVersion))
+	if err != nil {
+		allErrs = append(
+			allErrs,
+			field.Invalid(path, newWorkerVersion, fmt.Sprintf("could not parse version: %v, %v", oldWorkerVersion, err)))
+		return allErrs
+	}
+
+	parsedOldVersion, err := version.ParseGeneric(string(oldCPVersion))
+	if err != nil {
+		allErrs = append(
+			allErrs,
+			field.Invalid(path, newWorkerVersion, fmt.Sprintf("could not parse version: %v, %v", oldCPVersion, err)))
+		return allErrs
+	}
+
+	parsedNewVersion, err := version.ParseGeneric(string(newCPVersion))
+	if err != nil {
+		allErrs = append(
+			allErrs,
+			field.Invalid(path, newWorkerVersion, fmt.Sprintf("could not parse version: %v, %v", newCPVersion, err)))
+		return allErrs
+	}
+
+	if parsedOldVersion.LessThan(parsedNewVersion) {
+		allErrs = append(
+			allErrs,
+			field.Invalid(path, newWorkerVersion, fmt.Sprintf("can't simultaneously remove worker kubernetesVersion and upgrade top level kubernetesVersion: %v", newCPVersion)))
```

Comment:
What's the reason for not allowing removing and upgrading the top version simultaneous? Can't we treat a removal like an upgrade to the top version?

Reply:
We don't want to skip minor versions. If the old worker node version is at 1.19 and the old cluster version is 1.20 but the new cluster version is 1.21, then the worker node will try to jump from 1.19 -> 1.21 which has a skew of two minor versions. 

It's fine if the CP node is two minor versions above Worker Nodes, but we still want to upgrade nodes one minor version at a time.